pkg/engine/renderer/dependencies.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	return de.cache[d], nil
+}
+
+// sanitizeAndSerialize removes volatile parts of an object and returns the resulting object as serialized yaml
+func sanitizeAndSerialize(obj metav1.Object) (string, error) {
+	// Namespace is ignored mostly to allow easier integration tests (which use random namespaces)
+	ns := obj.GetNamespace()
+
+	// OwnerReferences need to be skipped as they contain a changing UID
+	or := obj.GetOwnerReferences()
+
+	// Annotations are notorious for containing quickly changing strings: plan/phase/task names, uids, dates, etc.
+	ann := obj.GetAnnotations()
+
+	obj.SetNamespace("")
+	obj.SetOwnerReferences([]metav1.OwnerReference{})
+	obj.SetAnnotations(map[string]string{})
+	yamlStr, err := ToYaml(obj)
+	obj.SetNamespace(ns)
+	obj.SetOwnerReferences(or)
+	obj.SetAnnotations(ann)
+	return yamlStr, err
+}
+
+// resourceDependency returns the resource of type t with the given namespace/name, either from the passed in list of objects or the last applied configuration from the API server
+func (de *dependencyCalculator) resourceDependency(d resourceDependency) (*unstructured.Unstructured, error) {
+
+	// First try to find the dependency in the local list, if it's deployed in the same task we'll find it here
+	for _, obj := range de.taskObjects {
+		if obj.GetObjectKind().GroupVersionKind().String() == d.gvk.String() {
+			if obj.GetName() == d.name && obj.GetNamespace() == d.namespace {
+				return obj, nil
+			}
+		}
+	}
+
+	// We haven't found it, so we need to query the api server to get the current version
+	//dep, _ := reflect.New(t).Interface().(metav1.Object)
+	dep := &unstructured.Unstructured{}
+	dep.SetGroupVersionKind(d.gvk)
+
+	key := client.ObjectKey{
+		Namespace: d.namespace,
+		Name:      d.name,
+	}
+
+	err := de.Client.Get(context.TODO(), key, dep)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve object %s/%s: %v", d.namespace, d.name, err)
+	}
+
+	// We don't want the hash from the object itself, because of added metadata from the api-server
+	// We use the LastAppliedConfigAnnotation that stores exactly what we applied last time
+	lastConfiguration, ok := dep.GetAnnotations()[kudo.LastAppliedConfigAnnotation]
+	if !ok {
+		return nil, fmt.Errorf("LastAppliedConfigAnnotation is not available on %s/%s", d.namespace, d.name)
+	}
+
+	obj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, []byte(lastConfiguration))
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode lastAppliedConfigAnnotation from %s/%s: %v", d.namespace, d.name, err)
+	}
+
+	return obj.(*unstructured.Unstructured), nil
+}
+
+// calculateResourceDependencies gets the resource dependencies of the passed in object
+func calculateResourceDependencies(obj *unstructured.Unstructured) (resourceDependencies, error) {
+	deps := resourceDependencies{}
+
+	// We can't rely on actual types here, so we just look into each possible path and try to find a pod template spec
```

Comment:
This way is fine too, but you could switch-case on `obj.GroupVersionKind`, right?

Reply:
It would have to be `obj.GroupKind` without the version, but yes, that would be possible.

I think it's ok like this for now.

At some point it probably would make sense to add companion types to all k8s types, that can:
- Check health
- Add correct labels & attributes
- Calculate resource dependencies
- Add dependencies hash
- ...

regardless of the actual api version